src/IRMatch.cpp

```diff
@@ -283,4 +283,126 @@ bool expr_match(Expr pattern, Expr expr, map<string, Expr> &matches) {
     }
 }
 
+namespace IRMatcher {
+
+HALIDE_ALWAYS_INLINE
+bool equal_helper(const Expr &a, const Expr &b) {
+    return equal(*a.get(), *b.get());
+}
+
+template<typename Op>
+HALIDE_ALWAYS_INLINE
+bool equal_helper_binop(const BaseExprNode &a, const BaseExprNode &b) {
+    return (equal_helper(((const Op &)a).a, ((const Op &)b).a) &&
+            equal_helper(((const Op &)a).b, ((const Op &)b).b));
+}
+
+HALIDE_ALWAYS_INLINE
+bool equal_helper(int a, int b) {
+    return a == b;
+}
+
+template<typename T>
+HALIDE_ALWAYS_INLINE
+bool equal_helper(const std::vector<T> &a, const std::vector<T> &b) {
+    if (a.size() != b.size()) return false;
+    for (size_t i = 0; i < a.size(); i++) {
+        if (!equal_helper(a[i], b[i])) return false;
+    }
+    return true;
+}
+
+bool equal_helper(const BaseExprNode &a, const BaseExprNode &b) {
```

Comment:
Is this all to avoid instantiating an IREquality visitor + virtual call? (comment)